setup.py

```diff
@@ -137,6 +137,8 @@ def finalize_options(self):
         'pybind11>=1.7',
         'psutil',
         'nibabel',
+        'typing',
+        'tensorflow<=1.15',  # brainiak.matnormal not compatible with tf2.0
```

Comment:
@mshvartsman, is there any chance to update the PR to use `tf.compat.v1` APIs such that it can be used with TensorFlow 2 without refactoring?

Reply:
At least some functionality didn't make it to tf.compat.v1 (e.g. inverse wishart and gamma distrs we use for priors), but let me assess how much work it'd be to reimplement or remove, stay tuned!

Reply:
Ok, here are the things without compat APIs as far as I can tell: 
1. Inverse-Wishart and Gamma priors for regularized covariances. I think we could possibly just remove those priors (I believe we used neither of them in the paper and it's probably not worth reimplementing over just moving to TF probability when we can). 
2. The `ScipyOptimizerInterface` we use to our problems into Scipy's solvers (usually L-BFGS) while handling session bookkeeping. Can't just cut this out -- can switch to something like Adam which would likely perform worse, or manage our own bookkeeping, or something else? Would appreciate advice on how to proceed here.

Reply:
Thanks for the quick investigation, @mshvartsman. Based on your findings, I think we should make compatibility with TensorFlow 2 a goal for merging the PR. I will try to get reviewers looking at it ASAP.

Reply:
(1) above is done. (2) has different options. The basic issue is that with most TF optimizers you can just pass a list of tensors and a loss and it knows how to optimize that loss w.r.t. those params -- this is how all the regular DL optimizers work (Adam, RMSProp, etc) and how `ScipyOptimizerInterface` worked in TF 1. For `scipy.optimize.minimize` and `tfp.optimizer.lbfgs_minimize` you need a function that takes theta and returns val and grad (f(theta) and f'(theta)). The current matrix-normal API doesn't really expect that, so the parameters are part of the various objects' states (model, cov etc) and spread out in random parts of the graph (e.g. structured covariance params, weights, etc). I see some options: 

1. Switch to one of the DL optimizers like Adam. It's a relatively easy change, but the optimizer will be worse than L-BFGS for our setting (no batches, no stochastic gradients). 
2. Write a bunch of kludgy code to take theta, unpack it, and inject it into the graph via `feed_dict` (if that's still a thing in TF 2). I think it'll work, but I'll need to be careful to make sure this always works no matter how people set up their models (it'd basically replicate the code in the old module I was using, https://github.com/tensorflow/tensorflow/blob/r1.15/tensorflow/contrib/opt/python/training/external_optimizer.py). 
3. Completely rewrite the API to be fully functional, so parameters will stop being parts of model state and we can just unpack them and pass them in. 

I don't like any of these options: are there any others? I'm not really a software guy :) so maybe there's something better to do here. Regardless, I'd like to get feedback from whoever will be reviewing (@mihaic / @lcnature / @szorowi1) before I get working on this, so we are on the same page on how to proceed.

Reply:
Good to hear about (1)! As far as I understand, @szorowi1 will be reviewing shortly. Let's address (2) after we hear from him.

Reply:
Hi all, happy to help with review. I started looking through the code, but it's been long enough at this point I've somewhat lost the thread. Are there specific sections of the code I should be reviewing? Are there sections that have already been reviewed and I should leave be? And/or am most helpful be helping to address (2) above? Thanks!

Reply:
Thanks for having another look at the PR, @szorowi1. We do need help with (2), so if you have experience with the options proposed by @mshvartsman, please weigh in.

As for the review in general, please give your opinion in as much detail as you can, based on the time you can commit.

Reply:
Note that this was reviewed by @lcnature already, and I believe I have addressed all his comments, so hopefully should be close with the exception of the remaining TF2.0 changes.

Reply:
I think I have the TF 2.0 workaround for the above that does not change any of the APIs and works with scipy.minimize: https://github.com/mshvartsman/brainiak/tree/mnorm-eager. My plan is to resolve @lcnature's comments in this PR first, then get the TF 2.0 fixes in. I don't know if there's a way to stack PRs in github: should I just merge that branch into this one, or do something else?

Reply:
Yes, I would prefer not to have a TF<2 dependency even temporarily, so please add those changes to this PR.

Reply:
Deleted previous comment, TF2.0 fix is still broken (I thought I had a clever solution but instead discovered how loose some of my tests are).